hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/upgrade/FinalizeScmUpgradeSubcommand.java

```diff
@@ -0,0 +1,152 @@
+import static org.apache.hadoop.hdds.scm.cli.upgrade.FinalizeUpgradeCommandUtil.isStarting;
+
+import java.io.IOException;
+import java.util.UUID;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+import org.apache.hadoop.hdds.cli.HddsVersionProvider;
+import org.apache.hadoop.hdds.scm.cli.ScmSubcommand;
+import org.apache.hadoop.hdds.scm.client.ScmClient;
+import org.apache.hadoop.ozone.upgrade.UpgradeException;
+import org.apache.hadoop.ozone.upgrade.UpgradeFinalizer;
+import org.apache.hadoop.ozone.upgrade.UpgradeFinalizer.StatusAndMessages;
+
+import picocli.CommandLine;
+
+/**
+ * Handler of Finalize SCM command.
+ */
+@CommandLine.Command(
+    name = "finalizeupgrade",
+    description = "Finalize SCM Upgrade",
+    mixinStandardHelpOptions = true,
+    versionProvider = HddsVersionProvider.class)
+
+public class FinalizeScmUpgradeSubcommand extends ScmSubcommand {
+  @CommandLine.Option(
+      names = {"--takeover"},
+      description = "Forces takeover of monitoring from another client, if "
+          + "finalization has already been started and did not finish yet."
+  )
+  private boolean force;
+
+  @Override
+  public void execute(ScmClient scmClient) throws IOException {
+    String upgradeClientID = "Upgrade-Client-" + UUID.randomUUID().toString();
+    try {
+      StatusAndMessages finalizationResponse =
+          scmClient.finalizeScmUpgrade(upgradeClientID);
+      if (isFinalized(finalizationResponse.status())){
+        System.out.println("Upgrade has already been finalized.");
+        emitExitMsg();
+        return;
+      } else if (!isStarting(finalizationResponse.status())){
+        System.err.println("Invalid response from Storage Container Manager.");
+        System.err.println(
+            "Current finalization status is: " + finalizationResponse.status()
+        );
+        throw new IOException("Exiting...");
+      }
+    } catch (UpgradeException e) {
+      handleInvalidRequestAfterInitiatingFinalization(force, e);
+    }
+    monitorAndWaitFinalization(scmClient, upgradeClientID);
+    return;
+  }
+
+  private void monitorAndWaitFinalization(ScmClient client,
+                                          String upgradeClientID)
+      throws IOException {
+    ExecutorService exec = Executors.newSingleThreadExecutor();
+    Future<?> monitor =
+        exec.submit(new UpgradeMonitor(client, upgradeClientID, force));
+    try {
+      monitor.get();
+      emitFinishedMsg("Storage Container Manager");
+    } catch (CancellationException |InterruptedException e) {
+      emitCancellationMsg("Storage Container Manager");
+    } catch (ExecutionException e) {
+      emitGeneralErrorMsg();
+      throw new IOException(e.getMessage());
```

Comment:
Thank you for making this change, one more thing still regarding this part, I think it is more beneficial to do not swallow the stack trace of this exception and having it reported in the output, so can we provide e.getCause() instead of e.getMessage() here to the wrapping IOException?

Reply:
Done.